themes/default/content/challenge/holiday-shopping.md

```diff
@@ -0,0 +1,269 @@
+    "grocery_frontend_image",
+    repository_url=repo.url,
+    path="./app/frontend")
+
+# Build and publish our application from /app/frontend and /app/backend as container images to the ECR repository
+backend_image = awsx.ecr.Image(
+    "grocery_backend_image",
+    repository_url=repo.url,
+    path="./app/backend")
+```
+
+Next, let's create an Application Load Balancer (ALB) to receive traffic and an [Amazon Elastic Container Service (ECS)](https://www.pulumi.com/learn/glossary/aws-ecs/) cluster in which the containers will run.
+
+```python
+# An ECS cluster to deploy into
+cluster = aws.ecs.Cluster("cluster")
+
+# An ALB to serve the frontend service to the internet
+lb = awsx.lb.ApplicationLoadBalancer("grocery-lb")
+```
+
+Now, create an [AWS Fargate](https://www.pulumi.com/learn/glossary/aws-fargate/) service that will run the containers. There are some configuration values for the Fargate service that you can set via Pulumi config, but for this example, you can just use the default values. This code creates a Fargate service that runs a task with two containers, a front and back. The front container is using the frontend image created in Step 1 and is wired to receive traffic from the ALB. The back container is using the backend image and listens on port 8000.
+
+```python
+# AWS configs
+container_port = config.get_int("containerPort", 80)
+cpu = config.get_int("cpu", 1024)
+memory = config.get_int("memory", 1024)
+
+# Deploy an ECS Service on Fargate to host the application containers
+service = awsx.ecs.FargateService(
+    "grocery-service",
+    cluster=cluster.arn,
+    task_definition_args=awsx.ecs.FargateServiceTaskDefinitionArgs(
+        containers={
+            "front": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
+                image=frontend_image.image_uri,
+                cpu=cpu,
+                memory=memory,
+                essential=True,
+                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
+                    container_port=container_port,
+                    target_group=lb.default_target_group,
+                )],
+            ),
+            "back": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
+                image=backend_image.image_uri,
+                cpu=cpu,
+                memory=memory,
+                essential=True,
+                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
+                    container_port=8000,
+                    host_port=8000
+                )],
+            ),
+        }  
+    ),
+    desired_count=1
+)
+
+# AWS export
+pulumi.export("app url", Output.concat("http://", lb.load_balancer.dns_name))
+```
+
+Run `pulumi up` and select yes to perform the update. Pulumi will output the `app url`. Go to the URL, and you will be presented with Pulumipus' Grocery List web application. You can try to add some items, but nothing will happen because there is no database connected yet. You will do that in the next step.
```

Comment:
This first deploy took a _really_ long time for me (the Fargate service is still `creating` after 10 minutes), so we should probably mention this up front to set expectations.

---

Update -- the `up` failed on this:

```
  aws:ecs:Service (grocery-service):
    error: 1 error occurred:
        * creating urn:pulumi:dev::atlas-fargate::awsx:ecs:FargateService$aws:ecs/service:Service::grocery-service: 1 error occurred:
        * error waiting for ECS service (arn:aws:ecs:us-west-2:616138583583:service/cluster-1529ef7/grocery-service-bd35866) to reach steady state after creation: ResourceNotReady: exceeded wait attempts
```

Health-check related, presumably?

Reply:
So this just takes like 6 minutes for me, but I added a note that it might take some time. I also don't get a 503 error

Reply:
Ok, I'll run through it again today then. (I actually don't think I've had a successful run of this at all on my own yet.)